### DISCUSSION TITLE
I create my own custom product. How to override existing product in the eCommerce?

### DISCUSSION THREAD

**Discussion by ferdware** [givanz/Vvveb#232] [General]
Hello Givanz,

So I created the structure of my own custom product. How do I override the existing product and use my own custom product?

My custom products uses variations different than the default one from the eCommerce. It also calculates the pricing variations differently.

Another alternative would be to use my custom product variations on the existing Product. How my custom product differs is that, the options create an exponential set of every combination of options. Then each option has it's own price and sku. 

Versus options just having a value that's added on top of the base value. 

Please Advise.

Thank You.

**Comment by ferdware** [comment]
Maybe this will provide better explanation. Here is an example of how Shopify creates variations and options. I want to achieve something similar to this.

![Screen Shot 2024-11-24 at 2 52 18 PM](https://github.com/user-attachments/assets/78939f10-388d-47cb-b552-1d2410f0894d)

**Comment by givanz** [comment]
Thanks for the suggestion, variations to override price and other options for product options combinations will be added after headless features are finished.

**Reply by ferdware** [reply]
Thank you for your reply. I'm going to see if I can build my own implementation. I do have a few questions.
1. Can I create my own custom product admin in my plugin? From what I understand I can.
2. Can I replace the existing **product.html** with my own from my "Plugin" or do I need to modify the existing **product.html** to add all my custom modifications?
Thank you again Givanz!

**Reply by ferdware** [reply]
I forgot the themes can be customized. Disregard queston #2.

**Reply by givanz** [reply]
Custom products can be created from admin https://demo.vvveb.com/admin/index.php?module=settings/product-types or from code with 

```php
Event::on('Vvveb\Controller\Base', 'customProduct', __CLASS__, function ($customProducts) {
	$customProducts += ['theme' => [
		'type'   => 'theme',
		'plural' => 'themes',
		//'icon-img'       => PUBLIC_PATH . 'plugins/theme-market/theme-market.svg',
		'icon'   => 'icon-copy-outline',
	]];

	return [$customProducts];
});
```

this example is from `theme-market` plugin that creates a `theme` custom product.

You can change product page template for a product type by using this code in your plugin

```php
function app() {
	
	Event::on('Vvveb\Controller\Product\Product', 'index', __CLASS__, function ($content, $language, $slug) {
		if ($content[$language]['type'] == 'product') {
			$content[$language]['template'] = 'product/my-template.html';
		}

		return [$content, $language, $slug];
	});		
}
```

**Reply by ferdware** [reply]
Thank You Givanz!

**Comment by meisbellajack** [comment]
If you don't want to write a custom pricing engine from scratch, you need a tool designed for Native Integration. This is exactly why we built [Misk Smart Product Options](https://apps.shopify.com/misk-smart-product-options).

Misk acts as the "Override" layer ferdware is looking for:

Bypasses the "Exponential" Limit: You can have 5 diameters, 10 widths, and 20 colors (1,000 variants). Misk handles the logic without you needing to manually create 1,000 rows in your Shopify admin.

Dynamic SKU/Price Generation: Just like the user asked, Misk can append or calculate the price and SKU based on the user's choices.

Example: Selecting Space Theme automatically adds $10 to the price and /SPACE to the SKU.

Native App Embed: Because it's built for the 2026 Shopify framework, it doesn't slow down your site or cause Cumulative Layout Shift (CLS), which keeps your SEO perfect.
